frontend: add file name to error messages

AlexanderViand opened the pull request:
This keeps track of the filename from which the frontend program is parsed, passes it around during the computation, and uses that to print a standard-style path/line/column in the error messages that should be clickable in most editors.

This is a PR into @georgemitenkov 's frontend-pr branch and shouldn't be merged into that until the current frontend PR (#376) on binary operations is merged.

I'm not super happy with the number of places where the filename needs to passed around. This seems to be necessary since the AST node objects apparently contain line and column, but not the filename. However, I might just be missing something.

georgemitenkov commented on xdsl/frontend/type_conversion.py:
```diff
@@ -39,6 +39,8 @@ class TypeConverter:
     what overloaded Python operations does this xDSL type support.
     """
 
+    file: str = field(default="<unknown file>")
```
Should this default to `None` instead?

georgemitenkov commented on xdsl/frontend/python_code_check.py:
```diff
@@ -258,11 +264,14 @@ class ConstantInliner(ast.NodeTransformer):
     new_node: ast.Constant
     """New AST node to inline."""
 
+    file: str = field(default="<unknown file>")
```
I guess here and in other places as well, I think defaulting to None is better.

georgemitenkov commented on xdsl/frontend/python_code_check.py:
```diff
@@ -45,29 +45,33 @@ def run(stmts: List[ast.stmt]) -> None:
 class SingleScopeVisitor(ast.NodeVisitor):
 
+    file: str = field(default="<unknown file>")
+    """File name for error reporting."""
```
nit: this can be a bit confusing in the future, because it's a path right?

AlexanderViand replied:
Good catch, thanks! The dangers of co-pilot autocomplete 🙈

georgemitenkov commented on xdsl/frontend/exception.py:
```diff
@@ -24,13 +24,21 @@ class CodeGenerationException(FrontendProgramException):
 
     def __str__(self) -> str:
-        return f"Code generation exception at {self.line}:{self.col}. {self.msg}"
+        return f'Code generation exception at "{self.file}", line {self.line} column {self.col}: {self.msg}'
```
We should probably check for None to avoid printing unknown_file/None

AlexanderViand replied:
Default changed & handling added

georgemitenkov commented on xdsl/frontend/code_generation.py:
```diff
@@ -251,8 +258,13 @@ def visit_FunctionDef(self, node: ast.FunctionDef) -> None:
     def visit_Name(self, node: ast.Name):
-        fetch_op = symref.Fetch.get(node.id, self.symbol_table[node.id])
-        self.inserter.insert_op(fetch_op)
+        try:
```
Can we have 
```
if node.id not in self.symbol_table:
  raise ...

fetch_op = symref.Fetch.get(node.id, self.symbol_table[node.id])
self.inserter.insert_op(fetch_op)
```

AlexanderViand replied:
Done!

georgemitenkov reviewed (approved):
Passing file everywhere is ugly, but I think we can keep it like this if you really want to see files in the errors already :)

btw, I think if you print error message like `file_path:line:col` then you can click it and go to that exact line (not super sure about that)

AlexanderViand commented:
Thanks for the feedback, I've addressed the comments. 

>btw, I think if you print error message like file_path:line:col then you can click it and go to that exact line (not super sure about that)
Python prints errors as, e.g., `"/home/alex/code/xdsl/xdsl/frontend/code_generation.py", line 65, in visit`, that's why I also went with the written out "line" and just added the column. The current format is picked up fine by vscode, but I'm happy to change it if other editors don't like it.

georgemitenkov commented:
@AlexanderViand should we merge this then?
